ospfd/ospf_auth.c

```diff
@@ -0,0 +1,707 @@
+	struct keychain *keychain = NULL;
+	struct key *key = NULL;
+	int key_id;
+
+	ibuf = STREAM_DATA(op->s);
+	ospfh = (struct ospf_header *)ibuf;
+
+	key_id = ospfh->u.crypt.key_id;
+
+	if (!OSPF_IF_PARAM(oi, keychain_name)) {
+		if (IS_DEBUG_OSPF_PACKET(ospfh->type - 1, RECV))
+			flog_warn(EC_OSPF_AUTH,
+				  "interface %s: Keychain is not set, Router-ID %pI4",
+				  IF_NAME(oi), &ospfh->router_id);
+		return 0;
+	}
+
+	keychain = oi->keychain;
+	if (!keychain) {
+		if (IS_DEBUG_OSPF_PACKET(ospfh->type - 1, SEND))
+			flog_warn(EC_OSPF_AUTH,
+				  "interface %s: Keychain %s is not available to send, Router-ID %pI4",
+				  IF_NAME(oi), OSPF_IF_PARAM(oi, keychain_name),
+				  &ospfh->router_id);
+		return 0;
+	}
+
+	key = oi->key;
+	if (!key) {
+		if (IS_DEBUG_OSPF_PACKET(ospfh->type - 1, SEND))
+			flog_warn(EC_OSPF_AUTH,
+				  "interface %s: Key ID %d not found in keychain %s, Router-ID %pI4",
+				  IF_NAME(oi), key_id, keychain->name,
+				  &ospfh->router_id);
+		return 0;
+	}
+
+	if (key->string == NULL || key->hash_algo == KEYCHAIN_ALGO_NULL) {
+		if (IS_DEBUG_OSPF_PACKET(ospfh->type - 1, SEND))
+			flog_warn(EC_OSPF_AUTH,
+				  "interface %s: Key ID %d in keychain %s is incomplete, Router-ID %pI4",
+				  IF_NAME(oi), key_id, keychain->name,
+				  &ospfh->router_id);
+		return 0;
+	}
+
+	/* Backward compatibility with RFC 2328 keyed-MD5 authentication */
+	if (key->hash_algo == KEYCHAIN_ALGO_MD5)
+		return ospf_auth_make_md5_digest(oi, op, key);
+	else
+		return ospf_auth_make_hmac_sha_digest(oi, op, key);
+}
+
+/* This function is called from ospf_write(), it will detect the
+ * authentication scheme and if it is MD5, it will change the sequence
+ * and update the MD5 digest.
+ */
+
+int ospf_auth_make(struct ospf_interface *oi,
+				struct ospf_packet *op)
+{
+	struct ospf_header *ospfh;
+	unsigned char digest[OSPF_AUTH_MD5_SIZE] = {0};
+#ifdef CRYPTO_OPENSSL
+	EVP_MD_CTX *ctx;
+#elif CRYPTO_INTERNAL
+	MD5_CTX ctx;
+#endif
+	void *ibuf;
+	uint32_t t;
+	struct crypt_key *ck;
+	const uint8_t *auth_key = NULL;
+
+	ibuf = STREAM_DATA(op->s);
+	ospfh = (struct ospf_header *)ibuf;
+
+	if (ntohs(ospfh->auth_type) != OSPF_AUTH_CRYPTOGRAPHIC)
+		return 0;
+
+	/* We do this here so when we dup a packet, we don't have to
+	 * waste CPU rewriting other headers.
+
+	   Note that frr_time /deliberately/ is not used here.
+	*/
+	t = (time(NULL) & 0xFFFFFFFF);
+	if (t > oi->crypt_seqnum)
+		oi->crypt_seqnum = t;
+	else
+		oi->crypt_seqnum++;
+
+	ospfh->u.crypt.crypt_seqnum = htonl(oi->crypt_seqnum);
+
+	/* Get MD5 Authentication key from auth_key list. */
+	if (list_isempty(OSPF_IF_PARAM(oi, auth_crypt)) && OSPF_IF_PARAM(oi, keychain_name) == NULL)
+		auth_key = (const uint8_t *)digest;
+	else if (!list_isempty(OSPF_IF_PARAM(oi, auth_crypt))) {
+		ck = listgetdata(listtail(OSPF_IF_PARAM(oi, auth_crypt)));
+		auth_key = ck->auth_key;
+	}
+
+	if (auth_key) {
+		/* Generate a digest for the entire packet + our secret key. */
+	#ifdef CRYPTO_OPENSSL
```

Comment:
Same here. `#ifdef` and related directives should start at the beginning of the line.